README
This archive contains the distribution App::perlall

Build, test and do with all perls

perlall -v help

    perlall build bleadd-nt
    perlall build perl5.16.2
    perlall build --allpatches perl5.14.2-nt
    perlall -v build -j4 5.15.4d-nt@khw-tk smoke-me/khw-tk
    perlall uninstall perl5.15.4d-nt@khw-tk

    perlall init perl5.16.2d-nt DBI CPAN::SQLite $(cat ~/Perl/B-C/t/top100)

    perlall do -MData::Dumper -e'my $a;$b={1=>\$a};$a=\$b;print Dumper($b)'
    perlall --older 5.12 make -Mblib t/0basic.t
    perlall=5.16.2  perlall maketest  	# test with version as ENV
    perlall="5.14*" perlall makeinstall
    perlall cpan My::Module
    perlall cpanm More::Modules

    perlall maketest "5.*.d*" # test with all debugging

    perlall initvm -v --all
    perlall testvm centos4 centos5 solaris10
    perlall testvm --all -c=list # see testvm in .perlall

Naming scheme: d DEBUGGING, -nt non-threaded, -m multi, @ git id

Configuration in source-able shell script ~/.perlall

You might want to link perlall to a valid command with "-command" appended
  perlall-make perlall-do perlall-cpan perlall-maketest...

See the presentation at https://docs.google.com/present/view?id=dcqmmdbh_4dmsxdqgx

I always disliked perlbrew. This uses perl installations as intended
by standard Configure.

This software is copyright (c) 2011-2013 by cPanel Inc.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
-*- change-log -*-

0.51    2019-12-10
        * build: cperl support

0.50    2016-06-12
        * build: cmdline options -D or -U override now default settings.
        * list: support cperl and perl

0.49    2015-11-27
        * Devel::PatchPerl::Plugin::Compiler: mandatory fix for 5.22
          which broke ByteLoader and thus the bytecode compiler.

0.48    2015-10-06
        * Devel::PatchPerl::Plugin::Compiler: updated RT#81332 (B) for 5.18,5.20,5.22
        * build: fixed detection of cmdline options -DAU
        * build: run TEST_JOBS=4 make test_harness with build
        * build: default to -Uuseshrplib on darwin with DEBUGGING
        * build: do not derive $Config{useshrplib} from default perl anymore

0.47    2015-07-08
	* fix -strong bug with clang, leave now -fstack-protector and
          -fstack-protector-strong.
        * add -ddflow, -sstack, -cps, -cpi suffix support for clang
        * remove support for old -faddress-sanitizer syntax

0.46    2014-08-09
	* fix -*san sanitizer cmdline quoting, add -isan support

0.45    2014-07-25
        * use nice _gitoneliner via git describe --long --tags --dirty --always

0.44    2014-01-11
        * build: fix -Uuseithreads with -nt suffix. it compiled to a threaded version since 0.42

0.43    2013-12-09
        * build: cp -rf usr/local/lib. libs are now installed readonly

0.42    2013-12-06
        * build: -m vs -mempool. do not build multi with 5.12.5d-mempool -Accflags=-DPERL_TRACK_MEMPOOL

0.41    2013-12-03
        * build: Collapse -thread-multi to -thread when only ithreads are available

0.40    2013-11-18
        * Devel::PatchPerl::Plugin::General: CPAN #88936 fixed with 5.19.6

0.39    2013-10-23
        * added new Devel::PatchPerl::Plugin::General for yet unpatched
          CPAN #88936 Compress-Raw-Zlib -g3: final link failed: Memory exhausted
          needed for debugging builds
        * Asan: fixed _patch_hsplit_rehash
        * fix typo in t/release-pod-syntax.t

0.38    2013-09-10
        * improved -asan detection with newer clang (> 3.2)

0.37    2013-07-13
        Asan:
        * improved RT#118525 patch "Return B::HEK for B::CV::GV of lexical subs"
          for 5.17-5.19

0.36    2013-07-11
        Asan:
        * add RT#118525 B::CV::GV crash on lexsubs for 5.17-5.19

0.35    2013-03-22
        Asan:
        * fix typo _patch_hsplit_rehash (#issue 6, Tokihiro Matsuno)

0.34    2013-03-05
        Asan:
        * improved RT#111594 Socket::unpack_sockaddr_un heap-buffer-overflow
          See https://rt.cpan.org/Public/Bug/Display.html?id=76067

0.33    2013-03-04
        Asan:
        * added fixes for CVE-2013-1667 hsplit rehash
        * fixed RT#115992 PL_eval_start use-after-free patch for 5.8.x

0.32    2013-02-23
        * skip installation on unsupported systems
        * skip tests as root
          cpantesters: http://analysis.cpantesters.org/solved?distv=App-perlall-0.31

0.31    2013-02-04
        * fix --allpatches and --patches=Compiler: disallow multiple
          Devel::PatchPerl Plugins, only one supported. Favor Asan over
          Compiler

0.30    2013-01-09
        * added lib/Devel/PatchPerl/Plugin/Compiler.pm
        * added option to build: --patches=name (multiply plugin name options allowed)
        * fixed local options, allow special case -v after make,do
        * improve maketest -v

0.29    2012-12-20
        * added to lib/Devel/PatchPerl/Plugin/Asan.pm:
          CPAN #82119 Socket::inet_ntop heap-buffer overflow
          RT#91678 S_anonymise_cv_maybe UTF8 cleanup
        * cmdline option: --allpatches (apply asan patches without asan)

0.28    2012-12-13
        * added to lib/Devel/PatchPerl/Plugin/Asan.pm:
          RT#115992 PL_eval_start use-after-free <5.15.3
          RT#115994 S_join_exact global-buffer-overflow 5.10-5.15.1

0.27    2012-11-15
        * mv scripts/perlall script/perlall to make it installable by EUMM
        * suffix -mad -> -Dmad

0.26    2012-11-13
        * hide Devel::PatchPerl from pause

0.25    2012-11-13
        * added to lib/Devel/PatchPerl/Plugin/Asan.pm:
          RT#115702 overlapping memcpy in to_utf8_case for 5.15.4-9, 5.17.0-6
          RT#111594 Socket::unpack_sockaddr_un heap-buffer-overflow <5.16.1
        * simplified patchlevel.h handling in patchperl
        * keep .git with --debug

0.24    2012-11-12
        * added lib/Devel/PatchPerl/Plugin/Asan.pm with
          RT#72700 List::Util Fix off-by-two on string literal length
          RT#111586 sdbm.c off-by-one access to global .dir
          for 5.10 - 5.14.
        * initvm installs Devel::PatchPerl::Plugin::Asan remotely

0.23    2012-11-11
        * fix -asan detection, fix -asan handling on darwin

0.22    2012-11-09
        * fix archname and startperl suffices. archname with cmdline suffix,
        startperl was broken since -Uversiononly
        * use -g -O1 -fsanitize=address for asan, clang 3.2svn required

0.21    2012-11-08
	* detect magic -clang, -asan and -tsan suffices

0.20    2012-11-08
	* fix option -l: do not print it double (hack)

0.19    2012-11-08
	* new option -l|--list for list

0.18    2012-11-06
	* set: new command to store the p alias in .perlall

0.17    2012-08-06
	* dynamic --main from git tag

0.16    2012-07-19
	* build 5.6: do not remove .git when we still need it (for the File::Find patch)
	* fix File::Find from 5.8.0 patch

0.15    2012-07-18
	* testvm: show the log results
	* dummy lib/App/perlall.pm added to be able to persuade 'cpan App::perlall'

0.14    2012-06-07
	* use default build useshrplib with clang asan (add -shared to lddlflags,
          force -Duseshrplib)
	* fix build 5.6.2 --install: missing sudo,
	  use File::Find from 5.8.0
	* do not download if perl-git is missing on build --install

0.13    2012-05-29  rurban
	* create seperate log.build-install not to overwrite log.build
	* fix system sudo cp * bindir/ on blead install

0.12    2012-05-05 rurban
	* fix INSTALL and perlall to work under ~ (mrmuskrat)

Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker 6.17; # perl-5.6.2
use Config;

die "unsupported OS $^O" # fixes welcome
  if $^O =~ /^vms|dos|bsdos$/;
die "$^O not yet fully supported\n" # fixes welcome
  if $^O =~ /^MSWin32|msys/;
die "Empty SHELL\n" # fail cpantesters
  unless $ENV{SHELL};

my %WriteMakefileArgs =
  (
   "NAME" => "App::perlall",
   "VERSION_FROM" => "script/perlall",
   "ABSTRACT" => "Build, test and do with all perls",
   "AUTHOR" => "Reini Urban <rurban\@cpanel.net>",
   "LICENSE" => "perl",
   'EXE_FILES'  => ['script/perlall'],
   "PREREQ_PM" =>  {
		    'App::Rad'     => "1.04",
		    'Devel::Platform::Info' => '0',
		    'Devel::PatchPerl' => "0.46",
		    'IO::Scalar'   => '0',
		    # in core:
		    'IPC::Cmd'     => '0.46', # in core since 5.10
		    'IPC::Open3'   => '0',
		    'Pod::Usage'   => "1.33",
		    'File::Basename' => "2.74",
		    'Cwd'          => "3.30",
		    'File::Path'   => "1.08",
		    'Getopt::Long' => "2.36",
		   },
   "SIGN" => 1,
   ($ExtUtils::MakeMaker::VERSION gt '6.46' ?
    ('META_MERGE'  =>
     {
      recommends =>
      {
       'File::Slurp' => "9999.13", # with B::C only
      },
      no_index =>
      {
	package     => 'B',
      },
      resources =>
      {
	license     => 'http://dev.perl.org/licenses/',
	bugtracker  => 'RT',
	repository  => 'git://github.com/rurban/App-perlall.git',
      }}) : ()),
  );

WriteMakefile(%WriteMakefileArgs);

if (! -f "$ENV{HOME}/.perlall") {
    my $cp = $Config{cp};
    system($cp, ".perlall", "$ENV{HOME}/.perlall");
    print <<INSTRUCTION;
Configuration in shell-syntax at ~/.perlall

You might want to add this to the end of your ~/.profile for
handy 'p', 'pm', 'mt', 'perl-git' and more aliases:

test -f ~/.perlall && . ~/.perlall
INSTRUCTION
}

package MY;
sub test {
    my $s = shift->SUPER::test(@_);
    # root can not run the tests on unix platforms
    $s =~ s/^(test_dynamic :: pure_all\n\t)(.+)$/$1if test `id -u` -gt 0; then $2; else echo skipped for root; true; fi/m
      if $^O =~ /^linux|freebsd|darwin|solaris|openbsd$/;
    $s;
}


# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 100
# End:
# vim: expandtab shiftwidth=4:


lib/Devel/PatchPerl/Plugin/General.pm
package Devel::PatchPerl::Plugin::General;
use base 'Devel::PatchPerl';

=head1 DESCRIPTION

Plugin for Devel::PatchPerl for general build patches which should
be provided by Devel::PatchPerl, but are not yet.

=head1 PATCHES

This list is complete:

    5.19.3-5: Compress::Raw::Zlib -g3 final link failed: Memory exhausted

=head2 Devel::PatchPerl::Plugin::General::patchperl($class, {version,source,patchexe})

Apply patches in Devel::PatchPerl::Plugin::General depending on the
perl version. See L<Devel::PatchPerl::Plugin>.

Every patch is recorded in patchlevel.h, visible in myconfig.
If a patch fails the script dies.

=cut

sub patchperl {
  my $class = shift;
  my %args = @_;
  my ($vers, $source, $patch_exe) = @args{qw(version source patchexe)};
  for my $p ( grep { Devel::PatchPerl::_is( $_->{perl}, $vers ) } @Devel::PatchPerl::patch ) {
    for my $s (@{$p->{subs}}) {
      my ($sub, @args) = @$s;
      push @args, $vers unless scalar @args;
      $sub->(@args);
    }
  }
}


package
  Devel::PatchPerl;

use File::Copy;
use vars '@patch';

@patch = (
  {
    perl => [ qr/^5\.19\.[345]$/ ],
    subs => [ [ \&_patch_CompressRawZlib] ],
  },
);

sub _add_patchlevel {
  my $vers = shift;
  my $line = shift;
  my $success;
  File::Copy::cp("patchlevel.h", "patchlevel.h.orig");
  open my $in, "<", "patchlevel.h.orig" or return;
  open my $out, ">", "patchlevel.h" or return;
  $line =~ s/"/\"/g;
  my $qr = $] > 5.010 ? /^\s+PERL_GIT_UNPUSHED_COMMITS/
                      : /^\tNULL$/;
  while (my $s = <$in>) {
    print $out $s;
    if ($s =~ $qr) {
      $success++;
      print $out "\t,\"".$line."\"\n";
    }
  }
  close $in;
  close $out;
  print STDERR "patched: $line\n";
  return $success;
}

sub _patch_CompressRawZlib
{
#From 82876fa94f7e69a7dc706d083a03f26b43d0cb4c Mon Sep 17 00:00:00 2001
#From: Reini Urban <rurban@x-ray.at>
#Date: Wed, 23 Oct 2013 10:38:28 -0500
#Subject: [PATCH] [CPAN #88936] Compress-Raw-Zlib -g3: final link failed:
# Memory exhausted
#
#---
# cpan/Compress-Raw-Zlib/Makefile.PL | 8 +++++++-
# 1 file changed, 7 insertions(+), 1 deletion(-)
  _patch(<<'END');
diff cpan/Compress-Raw-Zlib/Makefile.PL~ cpan/Compress-Raw-Zlib/Makefile.PL
index d8c060d..aba7abc 100644
--- cpan/Compress-Raw-Zlib/Makefile.PL~
+++ cpan/Compress-Raw-Zlib/Makefile.PL
@@ -6,6 +6,7 @@ require 5.006 ;
 use private::MakeUtil;
 use ExtUtils::MakeMaker 5.16 ;
 use ExtUtils::Install (); # only needed to check for version
+use Config;
 
 my $ZLIB_LIB ;
 my $ZLIB_INCLUDE ;
@@ -14,6 +15,10 @@ my $OLD_ZLIB = '' ;
 my $WALL = '' ;
 my $GZIP_OS_CODE = -1 ;
 my $USE_PPPORT_H = ($ENV{PERL_CORE}) ? '' : '-DUSE_PPPORT_H';
+my $OPTIMIZE = $Config{'optimize'};
+if ($Config{'gccversion'} and $OPTIMIZE =~ /-g3/) {
+  $OPTIMIZE =~ s/-g3/-g/; # [88936] out of memory with -g3 since 2.062
+}
 
 #$WALL = ' -pedantic ' if $Config{'cc'} =~ /gcc/ ;
 #$WALL = ' -Wall -Wno-comment ' if $Config{'cc'} =~ /gcc/ ;
@@ -81,7 +86,8 @@ WriteMakefile(
         ? zlib_files($ZLIB_LIB)
         : (LIBS => [ "-L$ZLIB_LIB -lz " ])
     ),
-      
+    OPTIMIZE => $OPTIMIZE,
+
     INSTALLDIRS => ($] >= 5.009 && $] < 5.011 ? 'perl' : 'site'),
 
     META_MERGE => {
-- 
1.8.4.rc3

END

  _add_patchlevel(@_, "CPAN #88936 Compress-Raw-Zlib -g3: final link failed: Memory exhausted");
}

1;


lib/Devel/PatchPerl/Plugin/Compiler.pm
package Devel::PatchPerl::Plugin::Compiler;
use base 'Devel::PatchPerl';

=head1 DESCRIPTION

The perl compiler modules L<B::C>, L<B::Bytecode> and L<B::CC> 
require various patches for various perl versions.

Some patches are mandatory for Windows or other strictly linked platforms (AIX), 
some are recommended to produce smaller and faster code.

You need to run C<perlall build --allpatches> or C<perlall build --patches=Compiler>
to apply these.

=head1 PATCHES

This list is complete for all perl versions 5.6 - 5.22.0, for all three threaded,
non-threaded and multiplicity variants.

    5.13.7-now:  RT#81332 revert 744aaba0 bloats the B compilers
    5.10-5.15.1: 8375c93e Export store_cop_label for the perl compiler
    5.15.2-3:    4497a11a Export DynaLoader symbols from libperl again
    5.22-        5.22 newPROG broke ByteLoader

=head2 Devel::PatchPerl::Plugin::Compiler::patchperl($class, {version,source,patchexe})

Apply patches in Devel::PatchPerl::Plugin::Compiler depending on the 
perl version. See L<Devel::PatchPerl::Plugin>

Every patch is recorded in patchlevel.h, visible in myconfig.
If a patch fails the script dies.

=cut

sub patchperl {
  my $class = shift;
  my %args = @_;
  my ($vers, $source, $patch_exe) = @args{qw(version source patchexe)};
  for my $p ( grep { Devel::PatchPerl::_is( $_->{perl}, $vers ) } @Devel::PatchPerl::patch ) {
    for my $s (@{$p->{subs}}) {
      my ($sub, @args) = @$s;
      push @args, $vers unless scalar @args;
      $sub->(@args);
    }
  }
}


package
  Devel::PatchPerl;

use File::Copy;
use vars '@patch';

@patch = (
  { perl => [ qr/^5\.1[456]\.\d$/ ],
    subs => [ [ \&_patch_B_BEGIN ] ],
  },
  { perl => [ qr/^5\.17\.\d$/ ], # TODO: 5.18,20,22
    subs => [ [ \&_patch_B_BEGIN_517 ] ],
  },
  { perl => [ qr/^5\.1[0-4]\.\d$/,
	      qr/^5\.15\.[01]$/ ],  # fixed in 5.15.2
    subs => [ [ \&_patch_store_cop_label] ],
  },
  { perl => [ qr/^5\.15\.[23]$/ ],
    subs => [ [ \&_patch_dl_export] ],
  },
  { perl => [ qr/^5\.2[23]/ ], # not yet fixed
    subs => [ [ \&_patch_522_ByteLoader] ],
  },
);

sub _add_patchlevel1 {
  my $vers = shift;
  my $line = shift;
  my $success;
  File::Copy::cp("patchlevel.h", "patchlevel.h.orig");
  open my $in, "<", "patchlevel.h.orig" or return;
  open my $out, ">", "patchlevel.h" or return;
  $line =~ s/"/\"/g;
  my $qr = $] > 5.010 ? /^\s+PERL_GIT_UNPUSHED_COMMITS/
                      : /^\tNULL$/;
  while (my $s = <$in>) {
    print $out $s;
    if ($s =~ $qr) {
      $success++;
      print $out "\t,\"".$line."\"\n";
    }
  }
  close $in;
  close $out;
  print STDERR "patch: $line\n";
  return $success;
}

sub _patch_B_BEGIN
{
  # Need to provide several versions per B version bump
  my $vers = shift;
  my $patch = <<'END';
diff -u ext/B/B.pm.orig ext/B/B.pm
--- ext/B/B.pm.orig	2012-05-23 17:40:02.167549196 -0500
+++ ext/B/B.pm	2013-01-09 13:15:10.391328942 -0600
@@ -6,26 +6,16 @@
 #      License or the Artistic License, as specified in the README file.
 #
 package B;
-use strict;
 
+require XSLoader;
 require Exporter;
 @B::ISA = qw(Exporter);
 
 # walkoptree_slow comes from B.pm (you are there),
 # walkoptree comes from B.xs
 
-BEGIN {
-    $B::VERSION = '1.34';
-    @B::EXPORT_OK = ();
-
-    # Our BOOT code needs $VERSION set, and will append to @EXPORT_OK.
-    # Want our constants loaded before the compiler meets OPf_KIDS below, as
-    # the combination of having the constant stay a Proxy Constant Subroutine
-    # and its value being inlined saves a little over .5K
-
-    require XSLoader;
-    XSLoader::load();
-}
+$B::VERSION = '1.34_01';
+@B::EXPORT_OK = ();
 
 push @B::EXPORT_OK, (qw(minus_c ppname save_BEGINs
 			class peekop cast_I32 cstring cchar hash threadsv_names
@@ -38,6 +28,9 @@
 			@specialsv_name
 		      ), $] > 5.009 && 'unitcheck_av');
 
+sub OPf_KIDS ();
+use strict;
+
 @B::SV::ISA = 'B::OBJECT';
 @B::NULL::ISA = 'B::SV';
 @B::PV::ISA = 'B::SV';
@@ -332,6 +325,8 @@
     }
 }
 
+XSLoader::load();
+
 1;
 
 __END__
END

  #; )
  # 5.14.0-3 B-1.29
  # 5.16.0   B-1.34
  # 5.16.1-2 B-1.35
  if ($vers =~ /^5\.14\./) {
      $patch =~ s/B::VERSION = '1.34/B::VERSION = '1.29/g;
  }
  elsif ($vers =~ /^5\.16\.[12]/) {
      $patch =~ s/B::VERSION = '1.34/B::VERSION = '1.35/g;
  }
  _patch($patch);

  _add_patchlevel1($vers, "RT#81332 revert 744aaba0 bloats the B compilers");
}

sub _patch_B_BEGIN_517
{
  # Need to provide several versions per B version bump
  my $vers = shift;
  my $patch = <<'END';
  _patch(<<'END');
diff -u ext/B/B.pm.orig ext/B/B.pm
--- ext/B/B.pm.orig	2012-11-28 16:28:25.376657707 -0600
+++ ext/B/B.pm	2013-01-09 13:35:28.247382145 -0600
@@ -6,26 +6,16 @@
 #      License or the Artistic License, as specified in the README file.
 #
 package B;
-use strict;
 
+require XSLoader;
 require Exporter;
 @B::ISA = qw(Exporter);
 
 # walkoptree_slow comes from B.pm (you are there),
 # walkoptree comes from B.xs
 
-BEGIN {
-    $B::VERSION = '1.41';
-    @B::EXPORT_OK = ();
-
-    # Our BOOT code needs $VERSION set, and will append to @EXPORT_OK.
-    # Want our constants loaded before the compiler meets OPf_KIDS below, as
-    # the combination of having the constant stay a Proxy Constant Subroutine
-    # and its value being inlined saves a little over .5K
-
-    require XSLoader;
-    XSLoader::load();
-}
+$B::VERSION = '1.41_01';
+@B::EXPORT_OK = ();
 
 push @B::EXPORT_OK, (qw(minus_c ppname save_BEGINs
 			class peekop cast_I32 cstring cchar hash threadsv_names
@@ -37,6 +27,9 @@
 			defstash curstash warnhook diehook inc_gv @optype
 			@specialsv_name unitcheck_av));
 
+sub OPf_KIDS ();
+use strict;
+
 @B::SV::ISA = 'B::OBJECT';
 @B::NULL::ISA = 'B::SV';
 @B::PV::ISA = 'B::SV';
@@ -330,6 +323,8 @@
     }
 }
 
+XSLoader::load();
+
 1;
 
 __END__
END

  #; )
  # 5.17.5 B-1.39
  # 5.17.6 B-1.40
  # 5.17.8 B-1.41
  if ($vers =~ /^5\.17\.5/) {
      $patch =~ s/B::VERSION = '1.41/B::VERSION = '1.39/g;
  }
  elsif ($vers =~ /^5\.17\.6/) {
      $patch =~ s/B::VERSION = '1.41/B::VERSION = '1.40/g;
  }
  elsif ($vers =~ /^5\.17\.[789]/) {
      $patch =~ s/B::VERSION = '1.41/B::VERSION = '1.41/g;
  }
  elsif ($vers =~ /^5\.18\.0/) {
      $patch =~ s/B::VERSION = '1.41/B::VERSION = '1.42/g;
  }
  elsif ($vers =~ /^5\.18\.1/) {
      $patch =~ s/B::VERSION = '1.41/B::VERSION = '1.42_01/g;
  }
  elsif ($vers =~ /^5\.18\.2/) {
      $patch =~ s/B::VERSION = '1.41/B::VERSION = '1.42_02/g;
  }
  elsif ($vers =~ /^5\.20\./) {
      $patch =~ s/B::VERSION = '1.41/B::VERSION = '1.48/g;
  }
  elsif ($vers =~ /^5\.22\./) {
      $patch =~ s/B::VERSION = '1.41/B::VERSION = '1.58/g;
  }
  _patch($patch);

  _add_patchlevel1($vers, "RT#81332 revert 744aaba0 bloats the B compilers");
}

sub _patch_store_cop_label
{
  # a70c7e2f048a735797bc368a5de5669f371e78fa Export store_cop_label for the perl compiler
  _patch(<<'END');
--- embed.fnc.old
+++ embed.fnc
@@ -2446,8 +2446,8 @@ Apon	|void	|sys_init3	|NN int* argc|NN char*** argv|NN char*** env
 Apon	|void	|sys_term
 ApoM	|const char *|fetch_cop_label|NN COP *const cop \
 		|NULLOK STRLEN *len|NULLOK U32 *flags
-: Only used  in op.c
-xpoM	|void|store_cop_label \
+: Only used  in op.c and the perl compiler
+ApoM	|void|store_cop_label \
 		|NN COP *const cop|NN const char *label|STRLEN len|U32 flags
 
 xpo	|int	|keyword_plugin_standard|NN char* keyword_ptr|STRLEN keyword_len|NN OP** op_ptr
diff --git a/ext/XS-APItest/APItest.xs b/ext/XS-APItest/APItest.xs
index 21f417d..6164bd0 100644
--- ext/XS-APItest/APItest.xs.old
+++ ext/XS-APItest/APItest.xs
@@ -2348,6 +2348,27 @@ test_cophh()
 #undef msvpvs
 #undef msviv
 
+void
+test_coplabel()
+    PREINIT:
+        COP *cop;
+        char *label;
+        int len, utf8;
+    CODE:
+        cop = &PL_compiling;
+        Perl_store_cop_label(aTHX_ cop, "foo", 3, 0);
+        label = Perl_fetch_cop_label(aTHX_ cop, &len, &utf8);
+        if (strcmp(label,"foo")) croak("fail # fetch_cop_label label");
+        if (len != 3) croak("fail # fetch_cop_label len");
+        if (utf8) croak("fail # fetch_cop_label utf8");
+        /* SMALL GERMAN UMLAUT A */
+        Perl_store_cop_label(aTHX_ cop, "foä", 4, SVf_UTF8);
+        label = Perl_fetch_cop_label(aTHX_ cop, &len, &utf8);
+        if (strcmp(label,"foä")) croak("fail # fetch_cop_label label");
+        if (len != 3) croak("fail # fetch_cop_label len");
+        if (!utf8) croak("fail # fetch_cop_label utf8");
+
+
 HV *
 example_cophh_2hv()
     PREINIT:
diff --git a/hv.c b/hv.c
index a230c16..11c5565 100644
--- hv.c.old
+++ hv.c
@@ -3286,6 +3286,15 @@ Perl_refcounted_he_inc(pTHX_ struct refcounted_he *he)
     return he;
 }
 
+/*
+=for apidoc fetch_cop_label
+
+Returns the label attached to a cop.
+The flags pointer may be set to C<SVf_UTF8> or 0.
+
+=cut
+*/
+
 /* pp_entereval is aware that labels are stored with a key ':' at the top of
    the linked list.  */
 const char *
@@ -3322,6 +3331,15 @@ Perl_fetch_cop_label(pTHX_ COP *const cop, STRLEN *len, U32 *flags) {
     return chain->refcounted_he_data + 1;
 }
 
+/*
+=for apidoc store_cop_label
+
+Save a label into a C<cop_hints_hash>. You need to set flags to C<SVf_UTF8>
+for a utf-8 label.
+
+=cut
+*/
+
 void
 Perl_store_cop_label(pTHX_ COP *const cop, const char *label, STRLEN len,
 		     U32 flags)
END

  _add_patchlevel1(@_, "a70c7e2f Export store_cop_label for the perl compiler");
}

sub _patch_dl_export
{
  # 5.15.2-3:    4497a11a Export DynaLoader symbols from libperl again
  _patch(<<'END');
diff --git a/ext/DynaLoader/dlutils.c b/ext/DynaLoader/dlutils.c
index 1ba9a61..574ccad 100644
--- ext/DynaLoader/dlutils.c.old
+++ ext/DynaLoader/dlutils.c
@@ -8,6 +8,7 @@
  *                      files when the interpreter exits
  */
 
+#define PERL_EUPXS_ALWAYS_EXPORT
 #ifndef START_MY_CXT /* Some IDEs try compiling this standalone. */
 #   include "EXTERN.h"
 #   include "perl.h"
END

  _add_patchlevel1(@_, "4497a11a Export DynaLoader symbols from libperl again");
}

sub _patch_522_ByteLoader
{
  # 5.22 broke ByteLoader
  _patch(<<'END');
index cab214a..a879e81 100644
--- op.c
+++ op.c
@@ -4058,7 +4058,9 @@ Perl_newPROG(pTHX_ OP *o)
 	PL_savestack_ix = i;
     }
     else {
-	if (o->op_type == OP_STUB) {
+	if (o->op_type == OP_STUB
+        || (o->op_type == OP_LINESEQ && OP_TYPE_IS(cUNOPo->op_first, OP_STUB)))
+        {
             /* This block is entered if nothing is compiled for the main
                program. This will be the case for an genuinely empty main
                program, or one which only has BEGIN blocks etc, so already
@@ -4069,9 +4071,8 @@ Perl_newPROG(pTHX_ OP *o)
                c71fccf11fde0068, changed perly.y so that newPROG() is now
                called with the output of block_end(), which returns a new
                OP_STUB for the case of an empty optree. ByteLoader (and
-               maybe other things) also take this path, because they set up
-               PL_main_start and PL_main_root directly, without generating an
-               optree.
+               PL_main_start and PL_main_root directly which should not be
+               overwritten by this empty PL_compcv.
 
                If the parsing the main program aborts (due to parse errors,
                or due to BEGIN or similar calling exit), then newPROG()
END

  _add_patchlevel1(@_, "5.22 newPROG broke ByteLoader");
}

1;
